src/redux/dispatch/DispatchBuilder.ts

```diff
@@ -0,0 +1,92 @@
+
+import { Reducer, Action, Store } from "redux";
+import { store } from "../../index.js";
+
+/**
+ * Helper class to make action dispatchers (self-dispatching action creator) and also reducers.
+ */
+export class DispatchBuilder<TState> {
+
+    /** 
+     * Initial text of all action types made by this builder.
+     * This helps generate unique action types across the application.
+     * The name of the state slice is a good suggestion.
+     */
+    actionNamePrefix: string;
+
+    /**
+     * All the actions that have been created by this builder. Used to make a reducer.
+     * The key is the action type.
+     * The value is a reducer function that supports that particular action.
+     * This can't be strongly typed, because each action potentially has a different payload type.
+     */
+    actionReducers: Map<string, Reducer<TState, ActionWithPayload<any>>>;
+
+    constructor(actionNamePrefix: string) {
+        this.actionNamePrefix = actionNamePrefix;
+        this.actionReducers = new Map<string, Reducer<TState, ActionWithPayload<any>>>();
+    }
+
+    /**
+     * Registers a new action. Returns a function which dispatches this action.
+     * @param actionName human-readable name for this action. Must be unique across this DispatchBuilder. For diagnostics (Redux debug) only.
+     * @param reducer The reducer function for this particular action. 
+     */
+    AddAction<TPayload>(actionName: string, reducer: (originalState: TState, payload: TPayload) => TState): (payload: TPayload) => void {
+
+        const actionType = `${this.actionNamePrefix} ${actionName}`;
+
+        // full reducer for just this action
+        const fullReducer: Reducer<TState, ActionWithPayload<TPayload>> = (state: TState, action: ActionWithPayload<TPayload>) => {
+
+            const payload = action.Payload;
+            const newState = reducer(state, payload);
+            return newState;
+        }
+
+        // save it!
+        this.actionReducers.set(actionType, fullReducer);
+
+        // action dispatcher = self-dispatching action creator
+        const actionDispatcher: (payload: TPayload) => void = (payload) => {
+
+            const action: ActionWithPayload<TPayload> = {
+                type: actionType,
+                Payload: payload,
+            };
+
+            const appStore = store as Store;
```

Comment:
This is importing the store from the JavaScript side, so it doesn't come typed.

Reply:
There is a compilation error as follows:

Type '(state: TState, action: ActionWithPayload<TPayload>) => TState' is not assignable to type 'Reducer<TState, ActionWithPayload<TPayload>>'.

Changing it to  (state: TState |undefined, action: ActionWithPayload<TPayload>) will probably fix up the error, but not sure whether it is your initial intension, or maybe error is caused by different version of Redux.

Reply:
The same compilation error happens to MakeReducer() as well, I understand this is a WIP, the logic of DispatchBuilder is very clear. It will be great if how to wire up AddAction() and Dispatch() from a component could be demonstrated.